### ISSUE TITLE
PhraseMatcher slower than Matcher when using attr="LOWER"?

### ISSUE THREAD

**Issue by fizban99** [explosion/spaCy#4109]
According to the [PhraseMatcher usage guide](https://spacy.io/usage/rule-based-matching#phrasematcher)

> If you need to match large terminology lists, you can also use the PhraseMatcher

and in the [Rule-based matching section](https://spacy.io/usage/rule-based-matching)

> The PhraseMatcher is useful if you already have a large terminology list or gazetteer consisting of single or multi-token phrases that you want to find exact instances of in your data.

and

> The Matcher isn’t as blazing fast as the PhraseMatcher

I was suprised at finding out that the Matcher seems to be significantly faster for single-term terminology lists when using "LOWER". Here is a sample code using a sample text from project Gutenberg and the results with %timeit:
```
import nltk
from nltk.corpus import gutenberg
from spacy.matcher import PhraseMatcher
from spacy.matcher import Matcher
from spacy.lang.en import English

nltk.download("gutenberg")
nlp = English()
doc = nlp(gutenberg.raw("carroll-alice.txt"))
words = ["alice", "she", "tired", "sitting", "sister"]

# with PhraseMatcher
matcher = PhraseMatcher(nlp.vocab, attr="LOWER")
terminology = list(nlp.tokenizer.pipe(words))
matcher.add("Test1", None, *terminology)
%timeit matches = matcher(doc)
# Result: 359 ms ± 55.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

# with Matcher
matcher = Matcher(nlp.vocab)
terminology = [{"LOWER": {"IN": words}}]
matcher.add("Test1", None, terminology)
%timeit matches = matcher(doc)
#Result: 65.1 ms ± 5.01 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
On the other hand, when not using "LOWER", the difference favours the PhraseMatcher:

```
import nltk
from nltk.corpus import gutenberg
from spacy.matcher import PhraseMatcher
from spacy.matcher import Matcher
from spacy.lang.en import English

nltk.download("gutenberg")
nlp = English()
doc = nlp(gutenberg.raw("carroll-alice.txt"))
words = ["alice", "she", "tired", "sitting", "sister"]

# with PhraseMatcher
matcher = PhraseMatcher(nlp.vocab)
terminology = list(nlp.tokenizer.pipe(words))
matcher.add("Test1", None, *terminology)
%timeit matches = matcher(doc)
# Result: 39.1 ms ± 979 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# with Matcher
matcher = Matcher(nlp.vocab)
terminology = [{"ORTH": {"IN": words}}]
matcher.add("Test2", None, terminology)
%timeit matches = matcher(doc)
#Result: 54.8 ms ± 1.97 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

I would suggest reviewing the usage guide about the benefits of the PhraseMatcher limitting it to multi-term terminology or ORTH values. Alternatively, review the code to see why the PhraseMather performs poorly when using LOWER...

**Comment by honnibal** [comment]
The matcher basically runs in time `O(N*M)`, for a document of `N` words and a pattern list of `M` entries. The phrase-matcher's average-case complexity is much better. The specifics depend on the pattern dictionary, but it's usually somewhere between `O(N)` and `O(N*log(M))`. This has a big impact when the phrase dictionaries are very big --- when you have large M. For instance, if you want to match a dictionary with every title in Wikipedia, the phrase-matcher will hugely out-perform the matcher. When the pattern list is four words long, it doesn't.

I would actually recommend you rethink your benchmarking methodology in general. Running a benchmark that executes in *milliseconds* is almost never helpful. You really want to set up a realistic workload, e.g. something that runs for at least a minute, making sure that the inputs actually reflect what you're doing.

**Comment by fizban99** [comment]
Thank you. Ines's answer is helpful indeed. Actually, I posted this issue because I identified it in one of my projects and I tried to create a reduced example here that showed the difference. 
I did some additional tests and it seems that the difference is significant enough only when matching a terminology list that are single terms and the PhraseMatcher is called with attr='LOWER'. Looking at the profiler, it seems that in these instances, the creation of the document is not that inexpensive compared to the cost of the rest of the function, making the PhraseMatcher slower than the normal Matcher. I was not able to find a combination of N and M where the PhraseMatcher was indeed faster in a scenario of single-worded terms and attr='LOWER'.
The PhraseMatcher really shines when using multi-word terminology. In this instance, the PhraseMatcher is a speed demon compared to the Matcher. Even when using attr='LOWER', if the Terminology list is big enough (maybe more than 1000 terms), it out-performs the Matcher.
I understand this issue is closed, and that is fine with me. I just wanted to highlight that in my project  It was more appropriate to use the Matcher instead of the PhraseMatcher and the reason was that my terminology list was single-worded and I was using LOWER.
